File: CodingTestPlatform/BOJ/2_Silver/4/10866/10866.py
# 덱
# https://www.acmicpc.net/problem/10866

from collections import deque
import sys
input = sys.stdin.readline


class Deque:
    def __init__(self):
        self.d = deque([])

    def push_front(self, v):
        self.d.appendleft(v)

    def push_back(self, v):
        self.d.append(v)

    def pop_front(self):
        return self.d.popleft() if len(self.d) else -1

    def pop_back(self):
        return self.d.pop() if len(self.d) else -1

    def size(self):
        return len(self.d)

    def empty(self):
        return 0 if len(self.d) else 1

    def front(self):
        return self.d[0] if len(self.d) else -1

    def back(self):
        return self.d[len(self.d) - 1] if len(self.d) else -1


d = Deque()
N = int(input().rstrip())

for i in range(N):
    c = input().split()
    if c[0] == 'push_front':
        d.push_front(c[1])
    elif c[0] == 'push_back':
        d.push_back(c[1])
    elif c[0] == 'pop_front':
        print(d.pop_front())
    elif c[0] == 'pop_back':
        print(d.pop_back())
    elif c[0] == 'size':
        print(d.size())
    elif c[0] == 'empty':
        print(d.empty())
    elif c[0] == 'front':
        print(d.front())
    elif c[0] == 'back':
        print(d.back())

File: CodingTestPlatform/BOJ/2_Silver/4/10866/10866.js
/**
 * 덱
 * Silver4
 * https://www.acmicpc.net/problem/10866
 */

class Node {
    constructor(val) {
        this.val = val;
        this.prev = null;
        this.next = null;
    }
}

class Deque {
    constructor() {
        this._head = null;
        this._tail = null;
        this._size = 0;
    }

    push_front(val) {
        const node = new Node(val);
        if (this._head === null) {
            this._head = node;
            this._tail = node;
        } else {
            this._head.prev = node;
            node.next = this._head;
            this._head = node;
        }        
        this._size++;
        return null;
    }

    push_back(val) {
        const node = new Node(val);
        if (this._tail === null) {
            this._head = node;
            this._tail = node;
        } else {
            node.prev = this._tail;
            this._tail.next = node;
            this._tail = node;
        }
        this._size++;
        return null;
    }

    pop_front() {
        if (this._head === null) return -1;
        const temp = this._head;
        const val = temp.val;
        this._head = temp.next;
        if (this._head === null) {
            this._tail = null;
        } else if (this._head === this._tail) {
            this._head.prev = null;
            this._tail.next = null;
        }
        this._size--;
        
        return val;
    }

    pop_back() {
        if (this._tail === null) return -1;
        const temp = this._tail;
        const val = temp.val;
        this._tail = this._tail.prev;
        if (this._tail === null) {
            this._head = null;
        } else if (this._head === this._tail) {
            this._head.prev = null;
            this._tail.next = null;
        }
        this._size--;
        
        return val;
    }

    size() {
        return this._size;
    }

    empty() {
        return this._size === 0 ? 1 : 0;
    }

    front() {
        return this._head ? this._head.val : -1;
    }

    back() {
        return this._tail ? this._tail.val : -1;
    }
}

const [N, ...input] = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');
const d = new Deque(), result = [];

input.forEach((v) => {
    const [command, num] = v.split(" ");
    const val = (num !== undefined) ? d[command](num) : d[command]();
    if (val !== null) result.push(val);
});

console.log(result.join('\n'));